react-app/src/library/services/headset.ts

```diff
@@ -0,0 +1,238 @@
+        this.application.hostedContext.isHosted() ? this.jabraNative : this.jabra
+        // this.jabra
+      );
+    }
+    implementations.push(this.plantronics);
+    implementations.push(this.sennheiser);
+
+    this._implementations = implementations;
+    return this._implementations;
+  }
+
+  getHeadSetEventsSubject = (): BehaviorSubject<ConsumedHeadsetEvents> => {
+    return this._headsetEvents$;
+  };
+
+  private subscribeToHeadsetEvents (implementation: VendorImplementation) {
+    implementation.on('deviceAnsweredCall', this.handleDeviceAnsweredCall.bind(this));
+    implementation.on('deviceRejectedCall', this.handleDeviceRejectedCall.bind(this));
+    implementation.on('deviceEndedCall', this.handleDeviceEndedCall.bind(this));
+    implementation.on('deviceMuteChanged', this.handleDeviceMuteStatusChanged.bind(this));
+    implementation.on('deviceHoldStatusChanged', this.handleDeviceHoldStatusChanged.bind(this));
+    implementation.on('deviceEventLogs', this.handleDeviceLogs.bind(this));
+    implementation.on('deviceConnectionStatusChanged', this.handleDeviceConnectionStatusChanged.bind(this));
+    implementation.on('webHidPermissionRequested' as any, (payload: any) => {
+      console.log('**** Debug Headset WebHID Event ****');
+      this._headsetEvents$.next({ event: 'webHidPermissionRequested' as any, payload })
+    });
+  }
+
+  activeMicChange(newMicLabel: string): void {
+    const implementation = this.implementations.find((implementation) => implementation.deviceLabelMatchesVendor(newMicLabel));
+    if (implementation) {
+      this.changeImplementation(implementation, newMicLabel);
+    } else if (this.selectedImplementation) {
+      this.selectedImplementation.disconnect();
+    }
+  }
+
+  async changeImplementation(implementation: VendorImplementation | null, deviceLabel: string): Promise<void> {
+    if (implementation === this.selectedImplementation) {
+      return;
+    }
+
+    if (this.selectedImplementation) {
+      this.selectedImplementation.disconnect();
+    }
+
+    this.selectedImplementation = implementation;
+
+    if (implementation) {
+      implementation.connect(deviceLabel);
+    }
+
+    this._headsetEvents$.next({ event: 'implementationChanged', payload: implementation});
+  }
+
+  // possible options: conversationId, contactName
+  incomingCall(callInfo: CallInfo, hasOtherActiveCalls?: boolean): Promise<any> {
+    this.logger.info('Inside incomingCall of headset library');
+    const service = this.selectedImplementation;
+    if (!service || !service.isConnected) {
+      this.logger.info('Headset: No vendor headset connected [incomingCall]');
+      return Promise.resolve();
+    }
+    return service.incomingCall(callInfo, hasOtherActiveCalls);
+  }
+
+  // possible options: conversationId, contactName
+  outgoingCall(callInfo: CallInfo): Promise<any> {
+    const service = this.selectedImplementation;
```

Comment:
NIT: there's a lot of repetition in these fn's. Might be be worth trying something like this:
```
private performActionIfConnected (actionName: string, perform: (impl: VendorImplementation) => Promise<any>) {
    const impl = this.selectedImplementation;
    if (!impl || !impl.isConnected) {
      this.logger.info(`Headset: No vendor headset connected [${actionName}]`);
      return Promise.resolve();
    }

    return perform(impl);
  }

  outgoingCall(callInfo: CallInfo): Promise<any> {
    return this.performActionIfConnected('outgoingCall', (implementation) => implementation.outgoingCall(callInfo));
  }
```